### PR TITLE
fix return type error of ceil/floor function

### PR DESCRIPTION
**Pull request by lkk2003rty**
#4136
only fix wrong return type && some flag error
TODO:
   - flag missing:   FLOOR(c_timestamp) need NOT_NULL flag
   - return length error

### CODE REVIEW

File: expression/builtin_math.go [hunk 1]
```diff
@@ -393,6 +393,7 @@ func (c *ceilFunctionClass) getFunction(args []Expression, ctx context.Context)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	fixFloorAndCeilReturnFlag(&bf, args[0])
```
Comment (zz-jason) [inline expression/builtin_math.go]: can we change the name `fixFloorAndCeilReturnFlag` and `fixFloorAndCeilType` to a more meaningful one ?

Comment (lkk2003rty) [inline expression/builtin_math.go]: such as corretXXX ？Any suggestions？

Comment (zz-jason) [inline expression/builtin_math.go]: how about:
- s/fixFloorAndCeilReturnFlag/setFlag4FloorAndCeil/
- s/fixFloorAndCeilType/getEvalTp4FloorAndCeil/

File: expression/builtin_math.go [hunk 2]
```diff
@@ -493,15 +494,14 @@ type floorFunctionClass struct {
 	baseFunctionClass
 }
 
-// fixFloorAndCeilType fixes the evalTp of FLOOR and CEIL.
-func fixFloorAndCeilType(arg Expression) (retTp, argTp evalTp) {
+// getEvalTp4FloorAndCeil fixes the evalTp of FLOOR and CEIL.
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: s/ fixes/ gets

File: expression/builtin_math.go [hunk 3]
```diff
@@ -493,15 +494,14 @@ type floorFunctionClass struct {
...
+func getEvalTp4FloorAndCeil(arg Expression) (retTp, argTp evalTp) {
 	fieldTp := arg.GetType()
 	switch arg.GetTypeClass() {
 	case types.ClassInt:
-		if !mysql.HasUnsignedFlag(fieldTp.Flag) {
-			retTp, argTp = tpInt, tpInt
-		} else {
-			retTp, argTp = tpDecimal, tpInt
+		retTp, argTp = tpInt, tpInt
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: why change this check?
mysql returns `type Longlong` when the arg is type `signed big int`.

File: expression/builtin_math.go (hunk 2 above)
Comment (zz-jason) [inline expression/builtin_math.go]: change comment

File: expression/builtin_math.go [hunk 4]
```diff
@@ -515,6 +515,15 @@ func fixFloorAndCeilType(arg Expression) (retTp, argTp evalTp) {
 	return
 }
 
+// setFlag4FloorAndCeil add return flag of FLOOR and CEIL
```
Comment (zz-jason) [inline expression/builtin_math.go]: ditto

File: plan/typeinfer_test.go [hunk 5]
```diff
@@ -349,21 +349,25 @@ func (s *testPlanSuite) createTestCase4MathFuncs() []typeInferTestCase {
 		{"cot(c_binary)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, mysql.MaxRealWidth, types.UnspecifiedLength},
 
 		{"floor(c_int)", mysql.TypeLonglong, charset.CharsetBin, mysql.BinaryFlag, 11, 0},
+		{"floor(c_bigint)", mysql.TypeNewDecimal, charset.CharsetBin, mysql.BinaryFlag, 21, 0},
```
Comment (XuHuaiyu) [inline plan/typeinfer_test.go]: add test for unsigned column

File: expression/builtin_math.go [hunk 6]
```diff
@@ -526,11 +535,12 @@ func (c *floorFunctionClass) getFunction(args []Expression, ctx context.Context)
 		return nil, errors.Trace(err)
 	}
 
-	retTp, argTp := fixFloorAndCeilType(args[0])
+	retTp, argTp := getEvalTp4FloorAndCeil(args[0])
 	bf, err = newBaseBuiltinFuncWithTp(args, ctx, retTp, argTp)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	setFlag4FloorAndCeil(&bf, args[0])
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: we can only check whether `mysql.HasBinaryFlag(args[0].Flag)` here,
and `set bf.tp.Flag |= mysql.BinaryFlag` when it's true.

the other code seem useless?

File: expression/builtin_math.go (hunk 3 above)
Comment (lkk2003rty) [inline expression/builtin_math.go]: because only when the argument is mysql.TypeLonglong(no matter whether signed or unsigned) need to return NEWDECIMAL

Comment (XuHuaiyu) [inline expression/builtin_math.go]: ok

File: expression/builtin_math.go [hunk 7]
```diff
@@ -515,6 +515,15 @@ func fixFloorAndCeilType(arg Expression) (retTp, argTp evalTp) {
 	return
 }
 
+// setFlag4FloorAndCeil sets return flag of FLOOR and CEIL
+func setFlag4FloorAndCeil(bf *baseBuiltinFunc, arg Expression) {
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: It seems that we do not need to pass bf here, we can pass bf.tp here

File: expression/builtin_math.go [hunk 8]
```diff
@@ -515,6 +515,15 @@ func fixFloorAndCeilType(arg Expression) (retTp, argTp evalTp) {
...
+	fieldTp := arg.GetType()
+	if (fieldTp.Tp == mysql.TypeLong || fieldTp.Tp == mysql.TypeNewDecimal) && mysql.HasUnsignedFlag(fieldTp.Flag) {
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: we can add a function `IsTypeInt` in util/types/etc.go,
and check `if (IsTypeInt(fieldTp.Tp) && fieldTp.Tp != mysql.TypeLonglong) || fieldTp.Tp == mysql.TypeNewDecimal) && xxx`  here

File: expression/builtin_math.go [hunk 9]
```diff
@@ -515,6 +515,15 @@ func fixFloorAndCeilType(arg Expression) (retTp, argTp evalTp) {
...
+		bf.tp.Flag |= mysql.UnsignedFlag
+	}
+	// TODO: when argument type is timestamp, add not full flag
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: add `.` at the end of this comment

File: expression/builtin_math.go (hunk 8 above)
Comment (lkk2003rty) [inline expression/builtin_math.go]: other function also need such function IsTypeInt? I think current condition maybe more clear in this scenario.

File: expression/builtin_math.go (hunk 6 above)
Comment (lkk2003rty) [inline expression/builtin_math.go]: the other code means?

File: expression/builtin_math.go (hunk 7 above)
Comment (lkk2003rty) [inline expression/builtin_math.go]: that's right, fix it.

**Comment by XuHuaiyu** [comment]
LGTM, PTAL @zz-jason

**Pull request merged**